datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java

```diff
@@ -345,6 +444,55 @@ public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<In
     return tasksToCleanUp;
   }
 
+  @Override
+  protected int constructExpectedNumberOfTasks(DatastreamGroup dg, List<String> instances,
+      Map<String, Set<DatastreamTask>> currentAssignmentCopy) {
+    // Count the number of tasks present for this datastream
+    Set<DatastreamTask> allAliveTasks = new HashSet<>();
+    for (String instance : instances) {
+      List<DatastreamTask> foundDatastreamTasks =
+          Optional.ofNullable(currentAssignmentCopy.get(instance)).map(c ->
+              c.stream().filter(x -> x.getTaskPrefix().equals(dg.getTaskPrefix()) && !allAliveTasks.contains(x))
```

Comment:
You don't really need allAliveTasks and additional check if x exists in allAliveTasks, since all the datastream groups are called only once. This will be a NO-OP. Also, you can directly use the `count() operator.

Reply:
Remember why we added this allAliveTasks check? Wasn't it because the same task may belong under two different instances (if we perform step 3 to load balance the tasks)? We'd add the task first, and if assignment is interrupted before we remove the task from the other instance, we find it in both? For this reason, I thought it was more correct to skip the duplicate tasks while counting. Is my understanding correct about this check?

Considering that we'll remove this code, I've left it as is and will clean it up as part of the ZK node persistence changes.